server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java

```diff
@@ -91,86 +92,162 @@ private static void deleteFileIfExists(Path stateLocation, Directory directory,
-                        // this is important since some of the XContentBuilders write bytes on close.
-                        // in order to write the footer we need to prevent closing the actual index input.
-                    }
-                })) {
+            deleteFileIfExists(stateLocation, directory, fileName);
+        } catch (IOException e) {
+            logger.trace("clean up failed {}", stateLocation.resolve(fileName));
+        }
+    }
+
+    private static void performDirectoryCleanup(Path stateLocation, Directory stateDir, String tmpFileName) {
+        deleteFileIgnoreExceptions(stateLocation, stateDir, tmpFileName);
+        IOUtils.closeWhileHandlingException(stateDir);
+    }
 
-                    builder.startObject();
-                    {
-                        toXContent(builder, state);
+    private Directory writeStateToFirstLocation(final T state, Path stateLocation, String tmpFileName)
+            throws WriteStateException {
+        try {
+            Directory stateDir = newDirectory(stateLocation);
+            try {
+                deleteFileIfExists(stateLocation, stateDir, tmpFileName);
+                try (IndexOutput out = stateDir.createOutput(tmpFileName, IOContext.DEFAULT)) {
+                    CodecUtil.writeHeader(out, STATE_FILE_CODEC, STATE_FILE_VERSION);
+                    out.writeInt(FORMAT.index());
+                    try (XContentBuilder builder = newXContentBuilder(FORMAT, new IndexOutputOutputStream(out) {
+                        @Override
+                        public void close() throws IOException {
+                            // this is important since some of the XContentBuilders write bytes on close.
+                            // in order to write the footer we need to prevent closing the actual index input.
+                        }
+                    })) {
+
+                        builder.startObject();
+                        {
+                            toXContent(builder, state);
+                        }
+                        builder.endObject();
                     }
-                    builder.endObject();
+                    CodecUtil.writeFooter(out);
                 }
-                CodecUtil.writeFooter(out);
-            }
 
-            stateDir.sync(Collections.singleton(tmpFileName));
-            stateDir.rename(tmpFileName, fileName);
-            stateDir.syncMetaData();
-            logger.trace("written state to {}", stateLocation.resolve(fileName));
-        } finally {
-            deleteFileIfExists(stateLocation, stateDir, tmpFileName);
+                stateDir.sync(Collections.singleton(tmpFileName));
+            } catch (Exception e) {
+                // perform clean up only in case of exception, we need to keep directory open and temporary file on disk
+                // if everything is ok for the next algorithm steps
+                performDirectoryCleanup(stateLocation, stateDir, tmpFileName);
```

Comment:
Perhaps I'm missing something, but it looks like we do `performDirectoryCleanup` on every directory in a `finally` block in the top-level `write` method. Is this extra call also necessary?

Reply:
It's true that clean up is performed in the write method in finally block, but only for those directories that are added successfully to the list of directories. If this method throws an exception, directory won't be added to the list.

Reply:
I see. Could we just open all the directories up-front and then clean them all up at the end? If possible I think it's clearer that there are no leaks if you can see acquisition and release of resources to be paired up in a single method. It's not always possible, but here I think it is.

Reply:
Sure, that actually makes sense. Please see aa443d6fcda7b4e2a79f43dae672a9fed5f9e0fb